src/renderers/shared/stack/reconciler/ReactCompositeComponent.js

```diff
@@ -880,6 +886,12 @@ var ReactCompositeComponent = {
         'boolean value. Make sure to return true or false.',
         this.getName() || 'ReactCompositeComponent'
       );
+
+      warning(
+        shallowEqual(this._currentState, inst.state),
+        'this.state: Setting state directly with this.state is not ' +
+        'recommended. Instead, use this.setState().'
+      );
```

Comment:
That's fair. Adding the name of the component to the warning is relatively easy. I should have done that in the first place. Showing what was changed is definitely more difficult. If people think that's valuable then I'm willing to look into it.